### DISCUSSION TITLE
能否直接描述编译产物为 package

### DISCUSSION THREAD

**Discussion by kenchztk** [xmake-io/xmake#3070] [Q&A] [answered]
比如我们有自己的内部库，
使用上都是直接发布 .h 和 二进制编译产物.so或.lib，按照一定目录形式组织
想引入 xmake 编译系统，但是内部包不想对目录和直接二进制产物发布的形式做更改，
是否能在此基础上直接描述成package，有什么好的方案吗

**Comment by waruqi** [comment, marked as answer]
urls 直接改成二进制包，然后 on_install 走 os.cp 复制到指定目录就行了。。

参考 
https://github.com/xmake-io/xmake-repo/blob/ce72a2f24f83a6e064ce6e0fdadca014006eacb3/packages/l/libsdl_mixer/xmake.lua#L48

**Reply by kenchztk** [reply]
一定需要 urls 来下载吗，安装时可以跳过下载 urls 的动作吗

**Reply by augustheart** [reply]
> 一定需要 urls 来下载吗，安装时可以跳过下载 urls 的动作吗

可以用这个
set_sourcedir("$(projectdir)/third/wolfssl_5_5_3")

**Reply by waruqi** [reply]
二进制本地包参考。https://xmake.io/#/zh-cn/package/local_package?id=%e9%bb%98%e8%ae%a4%e6%89%93%e5%8c%85%e6%a0%bc%e5%bc%8f

**Reply by kenchztk** [reply]
如果package中有需要加入.cpp源代码，这是有办法实现的吗

**Reply by waruqi** [reply]
有源码肯定要编译，走 set_sourcedir 本地源码编译模式，https://xmake.io/#/package/local_3rd_source_library

**Reply by kenchztk** [reply]
使用 set_sourcedir 模式来支持现有的本地包目录，但是目前某个包就是夹杂着源码，暂时不希望编译成库
如果使用 package 描述之后，能否不手动在描述工程的时候额外将这个源码加入files，
而是在 package 域直接把这个添加额外源文件的工作给完成了？

**Reply by waruqi** [reply]
不能 你得自己配 或者自己写 rule

**Reply by kenchztk** [reply]
因为我本地的源码包已经携带了编译产物，源码包其实只包括头文件和lib
这种习惯应该是符合部分公司的编译方案

我描述完源码包，我不希望xmake再将这份数据拷贝到install目录
我先通过on_fetch让xmake找到include、links这些信息
再通过on_load 设置好链接、头文件路径

编译是成功的，但是 install target 的时候，已经不自动拷贝这个package了，
源码看不懂，但是好像跟install时的配置文件有些关系
如果需要拷贝我这样描述的package 需要我自己去写这个拷贝过程是吗

**Reply by waruqi** [reply]
> 因为我本地的源码包已经携带了编译产物，源码包其实只包括头文件和lib

只是头文件和 lib 就不算源码包，就是二进制库包，直接 on_install 走 os.cp copy安装就完事了。

> 我描述完源码包，我不希望xmake再将这份数据拷贝到install目录

为啥不要，tar、git 下载下来是缓存目录，之后会自动清理掉的，你直接 fetch 用上了，之后过几天被自动清理了，链接就失败了，只有 copy 到 install 下，才是固定的，又不会永久占两份磁盘空间

> 如果需要拷贝我这样描述的package 需要我自己去写这个拷贝过程是吗


参考现有包

https://github.com/xmake-io/xmake-repo/blob/2a69168819240892a4cc0967b8ecb7473af6ed49/packages/l/libsdl/xmake.lua#L135

**Reply by kenchztk** [reply]
怎么在 on_install 去拷贝我知道的，
但是我并没有用 xmake 辅助的下载源码包方式是因为我们内部直接使用 svn 做二进制库管理发布，这个习惯我是没法改变的
也就是说 svn 下载下来的目录可以认为是永久目录不是临时的，这个地方应该是不会被自动清理的吧

但是我又希望有 xmake 简洁的语法来描述工程，所以想使用 xmake
特别是 install target(不是库)时内部的脚本可以一同发布依赖 package
但是按我这种不 install 库到 xmake 管理的package目录，install target时发现不会一同安装依赖了

我知道安装target时一同发布依赖package这个过程我可以自己去在target域获取 pkg 信息来实现，
但是我在想是不是可以通过 on_install 时修改 installdir 的哪些配置文件
我就可以利用内置的脚本帮我实现install target时一同拷贝依赖，而不是自己再去写个脚本

**Reply by waruqi** [reply]
修改安装目录可以用 set_installdir ，或者可以直接用 on_fetch 具体用法，可以自己建个 static 工程，执行 `xmake package` 生成的本地二进制包参考下。。

**Reply by waruqi** [reply]
```lua
package("tbox")
    set_description("The tbox package")

    add_configs("shared", {description = "Build shared library.", default = false, type = "boolean", readonly = true})

    on_load(function (package)
        package:set("installdir", path.join(os.scriptdir(), package:plat(), package:arch(), package:mode()))
    end)

    on_fetch(function (package)
        local result = {}
        local libfiledir = (package:config("shared") and package:is_plat("windows", "mingw")) and "bin" or "lib"
        result.links = "tbox"
        result.linkdirs = package:installdir("lib")
        result.includedirs = package:installdir("include")
        result.libfiles = path.join(package:installdir(libfiledir), "libtbox.a")
        return result
    end)
```

**Reply by kenchztk** [reply]
是的，就是 libfiles 这个参数是我想要的...

**Reply by wuzhiguocarter** [reply]
rust的依赖描述、构建规则、安装打包都放到了一个Cargo.toml文件，后面xmake.lua会参考这种设计吗？
@waruqi

**Reply by wuzhiguocarter** [reply]
也就是说，内部项目的依赖描述、构建、打包的规则描述，直接从项目根目录里xmake.lua获取，而不是从xmake-repo里面获取。
当然，开源外部的依赖包还是走xmake-repo库的xmake.lua获取依赖拉取、构建、打包的规则。

@waruqi

**Reply by wuzhiguocarter** [reply]
如果项目自己根目录已经自带了xmake.lua，如果再写一份xmake.lua到自己的xmake-repo，感觉有一些冗余，对内部项目来说不是特别友好。

**Reply by waruqi** [reply]
那你直接把包配置写自己项目的 xmake.lua 里面不就好了么，写哪都行。。原本就支持

**Reply by wuzhiguocarter** [reply]
之前没有这么用过，谢谢了

**Reply by wuzhiguocarter** [reply]
还有一个疑问，如果包的配置也放到了项目根目录下的xmake.lua，依赖它的项目怎么描述对它的依赖呢？

**Reply by waruqi** [reply]
同样的用法 add_requires

**Reply by wuzhiguocarter** [reply]
非常感谢，问题已解决。
我使用了xmake package把所有内部库的产出物放到一个临时路径，然后主项目用add_repositories添加这个临时路径，add_require添加内部依赖，add_package引用内部依赖。

**Comment by waruqi** [comment]
> rust的依赖描述、构建规则、安装打包都放到了一个Cargo.toml文件，后面xmake.lua会参考这种设计吗？ @waruqi

? 现在不也放在 xmake.lua 么，还需要怎么搞？

**Reply by wuzhiguocarter** [reply]
学会使用了xmake package以及add_repositories之后，使用体验确实已经很接近cargo了。
